src/System.Windows.Forms/src/System/Windows/Forms/Control.cs

```diff
@@ -12287,17 +12341,32 @@ private void WmDpiChangedBeforeParent(ref Message m)
                     // If it is a container control that inherit Font and is scaled by parent, we simply scale Font
                     // and wait for OnFontChangedEvent caused by its parent. Otherwise, we scale Font and trigger
                     // 'OnFontChanged' event explicitly. ex: winforms designer in VS.
-                    if (TryGetExplicitlySetFont(out Font local) || this is not ContainerControl || !IsScaledByParent(this))
+
+                    var container = this as ContainerControl;
+                    if (TryGetExplicitlySetFont(out Font local) || container is null || !IsScaledByParent(this))
                     {
                         if (local is not null)
                         {
-                            Font = scaledFont;
+                            // Containers whose Font is set explicitly are replaced with ScaledFont according to new Dpi but,
+                            // we are avoiding OnFontChangedEvent (if their AutoscaleMode is Inherit) that could trigger
+                            // PerformAutoscale with parents AutoscaleFactor. Reason for this is, Parent not yet received
+                            // DpiChanged event and thus neither scaled nor updated its AutoscaleFactor. Containers are marked
+                            // here as required scaling and will be scaled when the parent updated its AutoscaleFactor and ready to scale.
```

Comment:
Suggest to reword (hopefully I got the meaning correctly):
```suggestion
                            // Container controls with the font set explicitly have their fonts scaled according to the current dpi (i.e. set to ScaledFont).
                            // For container control with  AutoscaleMode=Inherit we'd like to avoid raising OnFontChanged event that would trigger
                            // PerformAutoscale with the parent's AutoscaleFactor. This is because the parent control is yet to receive DpiChanged event
                            // and, thus, neither scaled nor updated its AutoscaleFactor value.
                            // Mark the containers as required scaling, so they are scaled when their parents update AutoscaleFactor value and ready to scale.
```